server/controllers/finance/reports/balance/index.js

```diff
@@ -78,57 +82,96 @@ function document(req, res, next) {
  * @description
  * Returns the balance of the accounts for a given period.
  *
- * TODO(@jniles) - use the currencyId to get an exchange rate.
+ * The idea is this:
+ *  First column set must contain the opening balance of the fiscal year, no
+ *  matter what.
+ *  Second column set must contain the movements up to and including the
+ *  selected month.
+ *  Third contains the sum of all up to that month.
  *
+ * TODO(@jniles) - use the currencyId to get an exchange rate.
  */
 function getBalanceSummary(periodId, currencyId, shouldPrune) {
   const getFiscalYearSQL = `
-    SELECT id, start_date, end_date, fiscal_year_id FROM period WHERE id = ?
+    SELECT p.id, p.start_date, p.end_date, p.fiscal_year_id, p.number,
+      fy.start_date AS fiscalYearStart, fy.end_date
+    FROM period p JOIN fiscal_year fy ON p.fiscal_year_id = fy.id
+    WHERE p.id = ?;
   `;
 
   const sql = `
     SELECT a.id, a.number, a.label, a.type_id, a.label, a.parent,
       a.type_id = ${TITLE_ID} AS isTitleAccount,
       IFNULL(s.before, 0) AS "before",
-      IFNULL(s.before_debit, 0) AS "before_debit",
-      IFNULL(s.before_credit, 0) AS "before_credit",
+      IF(s.before > 0, s.before, 0) before_debit,
+      IF(s.before < 0, ABS(s.before), 0) before_credit,
       IFNULL(s.during, 0) AS "during",
       IFNULL(s.during_debit, 0) AS "during_debit",
       IFNULL(s.during_credit, 0) AS "during_credit",
       IFNULL(s.after, 0) AS "after",
-      IFNULL(s.after_debit, 0) AS "after_debit",
-      IFNULL(s.after_credit, 0) AS "after_credit",
+      IF(s.after > 0, s.after, 0) after_debit,
+      IF(s.after < 0, ABS(s.after), 0) after_credit,
       ? AS currencyId
     FROM account AS a LEFT JOIN (
       SELECT pt.account_id,
-        IF(pt.period_id < ?, SUM(pt.debit - pt.credit), 0) AS "before",
-        IF(pt.period_id < ?, SUM(pt.debit), 0) AS "before_debit",
-        IF(pt.period_id < ?, SUM(pt.credit), 0) AS "before_credit",
-        IF(pt.period_id = ?, SUM(pt.debit - pt.credit), 0) AS "during",
-        IF(pt.period_id = ?, SUM(pt.debit), 0) AS "during_debit",
-        IF(pt.period_id = ?, SUM(pt.credit), 0) AS "during_credit",
-        IF(pt.period_id <= ?, SUM(pt.debit - pt.credit), 0) AS "after",
-        IF(pt.period_id <= ?, SUM(pt.debit), 0) AS "after_debit",
-        IF(pt.period_id <= ?, SUM(pt.credit), 0) AS "after_credit"
+        SUM(IF(p.number = 0, pt.debit - pt.credit, 0)) AS "before",
+        SUM(IF(p.number BETWEEN 1 AND ?, pt.debit - pt.credit, 0)) AS "during",
+        SUM(IF(p.number BETWEEN 1 AND ?, pt.debit, 0)) AS "during_debit",
+        SUM(IF(p.number BETWEEN 1 AND ?, pt.credit, 0)) AS "during_credit",
+        SUM(IF(p.number <= ?, pt.debit - pt.credit, 0)) AS "after"
       FROM period_total AS pt
+        JOIN period p ON pt.period_id = p.id
+        JOIN account ac ON pt.account_id = ac.id
       WHERE pt.fiscal_year_id = ?
+        AND ac.type_id NOT IN (${INCOME_ID}, ${EXPENSE_ID})
```

Comment:
Why do we have this restriction ? I was thinking that those values (incomes and expenses) become null automatically when data in period 0 of the fiscal year

Reply:
Yes and no - the opening balance value is always 0, but this is for the movements throughout the year.  That's one advantage - since we never pass a transaction to make them 0 during the fiscal year, it allows us to always generate income/expense reports without worrying that the values become 0.

I just talked to Patrick about this.  We should hide these transactions if the fiscal year is closed, but we should keep it if the fiscal year is still open.  So, I need to add that code.